include/sleip/dynamic_array.hpp

```diff
@@ -0,0 +1,508 @@
+#ifndef SLEIP_DYNAMIC_ARRAY_HPP_
+#define SLEIP_DYNAMIC_ARRAY_HPP_
+
+#include <boost/core/alloc_construct.hpp>
+#include <boost/core/empty_value.hpp>
+#include <boost/throw_exception.hpp>
+#include <boost/config.hpp>
+#include <boost/iterator/iterator_concepts.hpp>
+
+#include <cstddef>
+#include <iterator>
+#include <memory>
+#include <utility>
+#include <type_traits>
+
+namespace sleip
+{
+namespace detail
+{
+template <class A, class T, class I>
+inline void
+alloc_move_construct_n(A& a, T* p, std::size_t n, I b)
+{
+  ::boost::detail::alloc_destroyer<A, T> hold(a, p);
+  for (std::size_t& i = hold.size(); i < n; void(++i), void(++b)) {
+    std::allocator_traits<A>::construct(a, p + i, std::move_if_noexcept(*b));
+  }
+  hold.size() = 0;
+}
+} // namespace detail
+
+template <class T, class Allocator = std::allocator<T>>
+struct dynamic_array;
+
+template <class T, class Allocator>
+struct dynamic_array : boost::empty_value<Allocator, 0>
+{
+public:
+  using value_type             = T;
+  using allocator_type         = Allocator;
+  using size_type              = std::size_t;
+  using difference_type        = std::ptrdiff_t;
+  using reference              = value_type&;
+  using const_reference        = value_type const&;
+  using pointer                = value_type*;
+  using const_pointer          = value_type const*;
+  using iterator               = pointer;
+  using const_iterator         = const_pointer;
+  using reverse_iterator       = std::reverse_iterator<iterator>;
+  using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+  static_assert(std::is_object_v<value_type>, "Only support object types");
+  static_assert(std::is_same_v<typename allocator_type::value_type, value_type>,
+                "Allocator's value type must match container's");
+
+private:
+  struct dealloc
+  {
+    Allocator   alloc;
+    std::size_t size = 0;
+
+    dealloc(dealloc const&) = delete;
+
+    dealloc(dealloc&& other) noexcept
+      : alloc(std::move(other.get_allocator()))
+      , size{other.size}
+    {
+      other.size = 0;
+    };
+
+    dealloc(Allocator const& alloc_, std::size_t size_)
+      : alloc(alloc_)
+      , size{size_}
+    {
+    }
+
+    auto
+    operator=(dealloc const&) & -> dealloc& = delete;
+
+    auto
+      operator=(dealloc&& other) &
+      noexcept -> dealloc&
+    {
```

Comment:
There should be no reason for move-assigning this.

Reply:
Great catch, thank you. The class is much simpler now.